composer/callbacks/export_for_inference.py

```diff
@@ -0,0 +1,84 @@
+
+"""Callback to export model for inference."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any, Optional, Sequence, Union
+
+from composer.core import State
+from composer.core.callback import Callback
+from composer.loggers import Logger
+from composer.utils.inference import ExportFormat, Transform, export_for_inference
+from composer.utils.object_store import ObjectStore
+
+log = logging.getLogger(__name__)
+
+__all__ = ['ExportForInferenceCallback']
+
+
+class ExportForInferenceCallback(Callback):
+    """Callback to export model for inference.
+
+    Example:
+        .. doctest::
+
+            >>> from composer import Trainer
+            >>> from composer.callbacks import ExportForInferenceCallback
+            >>> # constructing trainer object with this callback
+            >>> trainer = Trainer(
+            ...     model=model,
+            ...     train_dataloader=train_dataloader,
+            ...     eval_dataloader=eval_dataloader,
+            ...     optimizers=optimizer,
+            ...     max_duration="1ep",
+            ...     callbacks=[ExportForInferenceCallback(save_format='torchscript',save_path='/tmp/model.pth')],
+            ... )
+
+    Args:
+        save_format (Union[str, ExportFormat]):  Format to export to. Either ``"torchscript"`` or ``"onnx"``.
+        save_path (str): The path for storing the exported model. It can be a path to a file on the local disk,
+            a URL, or if ``save_object_store`` is set, the object name
+            in a cloud bucket. For example, ``my_run/exported_model``.
+        save_object_store (ObjectStore, optional): If the ``save_path`` is in an object name in a cloud bucket
+            (i.e. AWS S3 or Google Cloud Storage), an instance of
+            :class:`~.ObjectStore` which will be used
+            to store the exported model. Set this to ``None`` if ``save_path`` is a local filepath.
+            (default: ``None``)
+        sample_input (Any, optional): Example model inputs used for tracing. This is needed for "onnx" export
+        transforms (Sequence[Transform], optional): transformations (usually optimizations) that should
+            be applied to the model. Each Transform should be a callable that takes a model and returns a modified model.
+    """
+
+    def __init__(
+        self,
+        save_format: Union[str, ExportFormat],
+        save_path: str,
+        save_object_store: Optional[ObjectStore] = None,
+        sample_input: Optional[Any] = None,
+        transforms: Optional[Sequence[Transform]] = None,
+    ):
+        self.save_format = save_format
+        self.save_path = save_path
+        self.save_object_store = save_object_store
+        self.sample_input = sample_input
+        self.transforms = transforms
+
+    def after_dataloader(self, state: State, logger: Logger) -> None:
+        del logger
+        if self.sample_input is None:
+            self.sample_input = state.batch
```

Comment:
If the batch is modified in place, then algs like progressive resizing could still be caught in the sample batch here. Should this be a deepcopy? That said, a deepcopy would have memory (and by extension, throughput) overhead, so not thrilled about that either. This concern could be mitigated by doing the deepcopy only once (e.g on the first batch)

Reply:
Since sample_input will only be None once, adding deepcopy() to line 72 won't be a perf issue.